packetbeat/protos/smtp/parser.go

```diff
@@ -0,0 +1,261 @@
+package smtp
+
+import (
+	"bytes"
+	"errors"
+	"io/ioutil"
+	"net/mail"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/streambuf"
+	"github.com/elastic/beats/packetbeat/protos/applayer"
+)
+
+type parser struct {
+	buf     streambuf.Buffer
+	config  *parserConfig
+	pub     *transPub
+	message *message
+	state   parseState
+
+	onMessage func(m *message) error
+}
+
+type parserConfig struct {
+	maxBytes int
+}
+
+type message struct {
+	applayer.Message
+
+	// Request
+	command common.NetString
+	param   common.NetString
+	// Request data payload
+	headers map[string]common.NetString
+	body    common.NetString
+
+	// Response
+	statusCode    int
+	statusPhrases []common.NetString
+
+	raw []byte
```

Comment:
consider `streambuf.Buffer` for the raw data. Supports append and implements reader interface, tracking amount of bytes processed.